subprojects/core-api/src/main/java/org/gradle/api/attributes/java/TargetJavaPlatform.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.attributes.java;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.attributes.Attribute;
+
+/**
+ * Represents the target platform of a Java library or platform. The target level is expected to correspond
+ * to a Java platform version number (integer). For example, "5" for Java 5, "8" for Java 8, or "11" for Java 11.
+ *
+ * @since 5.3
+ */
+@Incubating
+public interface TargetJavaPlatform {
+
+    /**
+     * The minimal target platform for a Java library. Any consumer below this version would not be able to consume it.
+     */
+    Attribute<Integer> MINIMAL_TARGET_PLATFORM_ATTRIBUTE = Attribute.of("org.gradle.java.min.platform", Integer.class);
```

Comment:
We should use `JavaVersion` as the type for this, I think.

Reply:
Actually we decided _not_ to do it for different reasons:

1. `JavaVersion` has a lot of history and is not super nice. Mixing `VERSION_1_8` with `VERSION_10`, used to support "unknown versions", ...
2. If we want it to be future proof (Gradle next supports, say, Java 13 but this is consumed in Gradle current which does not), then parsing would fail with an unknown enum value. Working around this means implementing a special case in attribute matching/serialization, and we don't see any advantage in doing this.
3. All we need is a platform version number, which is very easily expressed with an integer. Should Java change (again) version schemes in the future, we would still be able to have rules which understand a newer version number. Basically it's like an enum, but more lenient/future proof. It also maps nicely to what you pass to `--release` in compilation.